content/blog/2022-03-08-alive/index.md

```diff
@@ -0,0 +1,114 @@
+
+
+## Background
+At its core, Alive is a domain specific language for formally verifying peephole optimizations in LLVM. Empirically, it found bugs where the optimization could introduce additional undefined behavior in situations where the original code did not. For me, the main important pieces of background to understand were LLVM, Peephole Optimizations, SMT Solvers, and Undefined Behaviors in C/C++.
+
+### LLVM
+According to [the project site](https://llvm.org), LLVM is “a collection of modular and reusable compiler and toolchain technologies”. Basically, compilers have a lot of similar components – they need to split up the program’s control flow to define all the basic blocks, perform dead code elimination, rearrange loops, and many other optimizations. LLVM makes it easier to reuse and compose these components by creating an _intermediate representation_ which is much closer to being machine code, but still has enough complexity and expressiveness to be able to figure out and perform the optimizations on a representation which is not yet specific to a particular architecture, and which is easier to think about.
+
+This project has an active development community and people do contribute individual optimizations and pieces of code, so architecturally this seems to have been a great choice.
+
+### SMT Solvers
+SMT stands for Satisfiability Modulo Theory. Basically an SMT solver takes in a mathematical formula with many unknowns, and tries to figure out if the formula is satisfiable in the sense that there is some way to assign the variables such that the formula is true.
+
+For example, in Boolean Logic the formula `x or y` is satisfiable with `x=True`, but `x and not x` is not satisfiable for any `x`.
+
+Fascinatingly, despite the fact that SMT is NP-complete, computer scientists have been developing faster and faster SMT solvers, and using them in a variety of problems. While in general solving satisfiability problems requires brute force, in practice it appears that lovingly crafted (and ruthlessly empirically tested) heuristics are able to get better and better results on the problems that people feed to SMT solvers. However, that is not to say that an SMT solver is likely to solve your problem well – setting up problems for SMT solvers to solve them successfully is a bit of an art, and it is fortunate when researchers discover encodings that work for their problems. For example, in this paper the researchers discovered that a hand-written implementation of pointer allocation and accessing resulted in faster SMT solving than using the theory of arrays, even though they are theoretically equivalent.
+
+### Peephole Optimizations
+Peephole optimizations are small optimizations which look at a small number of instructions to find ways to do the same task but make them more computationally efficient. 
```

Comment:
Maybe one example would be helpful here. Like, `-x < 0` can be rewritten to `x > 0` or whatever.